File: backend_typescript/src/AgentTeam/IngestionSplitter.ts
/**
 * Faithful TypeScript mirror of src/AgentTeam/Services/IngestionSplitter.php.
 *
 * Pure recursive character text splitter — a port of langchain's
 * RecursiveCharacterTextSplitter with library defaults: keep_separator="start",
 * is_separator_regex=False, strip_whitespace=True, length_function=len (code
 * points). The chunking boundaries/overlap math must match the PHP byte-for-byte
 * because it determines exactly what text gets embedded.
 *
 * All methods are static, mirroring the PHP `final class`.
 */

const DEFAULT_SEPARATORS = ['\n\n', '\n', ' ', ''];

/** PHP trim(): strips " \t\n\r\0\x0B" from both ends. */
const PHP_WS = ' \t\n\r\0\x0B';
function phpTrim(s: string): string {
  let start = 0;
  let end = s.length;
  while (start < end && PHP_WS.indexOf(s[start]) !== -1) start++;
  while (end > start && PHP_WS.indexOf(s[end - 1]) !== -1) end--;
  return s.slice(start, end);
}

/**
 * PHP preg_quote($str, '/'): escape regex metacharacters plus the delimiter.
 * Applied only to the fixed DEFAULT_SEPARATORS in practice, but implemented
 * faithfully. The special set matches PHP's preg_quote.
 */
function pregQuote(str: string, delimiter = '/'): string {
  const special = '.\\+*?[^]$(){}=!<>|:-#';
  let out = '';
  for (const ch of str) {
    if (ch === '\0') {
      out += '\\000';
    } else if (special.indexOf(ch) !== -1 || ch === delimiter) {
      out += '\\' + ch;
    } else {
      out += ch;
    }
  }
  return out;
}

export class IngestionSplitter {
  /**
   * Split text into chunks. Mirrors recursiveSplit() in the PHP.
   * @throws Error when overlap > chunkSize
   */
  static recursiveSplit(text: string, chunkSize = 1000, overlap = 150): string[] {
    if (overlap > chunkSize) {
      throw new Error(
        `Got a larger chunk overlap (${overlap}) than chunk size (${chunkSize}), should be smaller.`
      );
    }
    return IngestionSplitter.splitText(text, DEFAULT_SEPARATORS, chunkSize, overlap);
  }

  /** Code-point length (matches Python len() on str / PHP mb_strlen). */
  private static len(s: string): number {
    let n = 0;
    // for..of iterates by Unicode code point, matching mb_strlen(..,'UTF-8').
    for (const _ of s) n++;
    return n;
  }

  /**
   * Mirrors _split_text_with_regex with keep_separator="start": each separator
   * is attached to the piece that FOLLOWS it. `sepEscaped` is an already-quoted
   * regex (or '' for character split).
   */
  private static splitWithRegex(text: string, sepEscaped: string): string[] {
    let splits: string[];
    if (sepEscaped !== '') {
      // PHP preg_split('/(sep)/u', text, -1, PREG_SPLIT_DELIM_CAPTURE): the
      // captured delimiters are interleaved in the result — JS split with a
      // capturing group does the same.
      let parts: string[];
      try {
        parts = text.split(new RegExp('(' + sepEscaped + ')', 'u'));
      } catch {
        parts = [text];
      }
      splits = [];
      const n = parts.length;
      for (let i = 1; i < n; i += 2) {
        splits.push(parts[i] + (parts[i + 1] !== undefined ? parts[i + 1] : ''));
      }
      if (parts.length % 2 === 0) {
        splits.push(parts[parts.length - 1]);
      }
      splits.unshift(parts[0]);
    } else {
      // mb_str_split($text, 1, 'UTF-8') — one entry per code point.
      splits = Array.from(text);
    }
    return splits.filter((s) => s !== '');
  }

  /** Mirrors _join_docs (separator-join + trim; '' → null). */
  private static joinDocs(docs: string[], separator: string): string | null {
    const text = phpTrim(docs.join(separator));
    return text === '' ? null : text;
  }

  /**
   * Mirrors TextSplitter._merge_splits (length_function = len). With the default
   * keep_separator the merge separator is '' (separator_len 0).
   */
  private static mergeSplits(
    splits: string[],
    separator: string,
    chunkSize: number,
    chunkOverlap: number
  ): string[] {
    const separatorLen = IngestionSplitter.len(separator);
    const docs: string[] = [];
    const currentDoc: string[] = [];
    let total = 0;

    for (const d of splits) {
      const lenD = IngestionSplitter.len(d);
      if (total + lenD + (currentDoc.length > 0 ? separatorLen : 0) > chunkSize) {
        if (currentDoc.length > 0) {
          const doc = IngestionSplitter.joinDocs(currentDoc, separator);
          if (doc !== null) {
            docs.push(doc);
          }
          while (
            total > chunkOverlap ||
            (total + lenD + (currentDoc.length > 0 ? separatorLen : 0) > chunkSize && total > 0)
          ) {
            total -= IngestionSplitter.len(currentDoc[0]) + (currentDoc.length > 1 ? separatorLen : 0);
            currentDoc.shift();
          }
        }
      }
      currentDoc.push(d);
      total += lenD + (currentDoc.length > 1 ? separatorLen : 0);
    }

    const doc = IngestionSplitter.joinDocs(currentDoc, separator);
    if (doc !== null) {
      docs.push(doc);
    }
    return docs;
  }

  /** Mirrors RecursiveCharacterTextSplitter._split_text (is_separator_regex=False). */
  private static splitText(
    text: string,
    separators: string[],
    chunkSize: number,
    chunkOverlap: number
  ): string[] {
    const finalChunks: string[] = [];

    // Choose the separator: the last, unless an earlier one occurs in text.
    let separator = separators[separators.length - 1];
    let newSeparators: string[] = [];
    for (let i = 0; i < separators.length; i++) {
      const s = separators[i];
      const sEscaped = pregQuote(s, '/');
      if (s === '') {
        separator = s;
        break;
      }
      let matched = false;
      try {
        matched = new RegExp(sEscaped, 'u').test(text);
      } catch {
        matched = false;
      }
      if (matched) {
        separator = s;
        newSeparators = separators.slice(i + 1);
        break;
      }
    }

    const sepEscaped = pregQuote(separator, '/');
    const splits = IngestionSplitter.splitWithRegex(text, sepEscaped);

    // Merge short pieces; recurse into pieces still longer than chunk_size.
    let goodSplits: string[] = [];
    const mergeSeparator = ''; // keep_separator => merge with ''
    for (const s of splits) {
      if (IngestionSplitter.len(s) < chunkSize) {
        goodSplits.push(s);
      } else {
        if (goodSplits.length > 0) {
          const merged = IngestionSplitter.mergeSplits(goodSplits, mergeSeparator, chunkSize, chunkOverlap);
          finalChunks.push(...merged);
          goodSplits = [];
        }
        if (newSeparators.length === 0) {
          finalChunks.push(s);
        } else {
          const other = IngestionSplitter.splitText(s, newSeparators, chunkSize, chunkOverlap);
          finalChunks.push(...other);
        }
      }
    }
    if (goodSplits.length > 0) {
      const merged = IngestionSplitter.mergeSplits(goodSplits, mergeSeparator, chunkSize, chunkOverlap);
      finalChunks.push(...merged);
    }
    return finalChunks;
  }
}

File: langchain_runner/ingestion_splitter.py
"""Pure-Python recursive character text splitter.

Vendored from `langchain-text-splitters` (MIT License) -- the split ALGORITHM
ONLY, transcribed so it runs with no langchain / langchain_core /
langchain_text_splitters runtime dependency. This lets the same splitting logic
run under Pyodide (browser Python), where langchain cannot load.

Output is intended to match
``RecursiveCharacterTextSplitter(chunk_size, chunk_overlap).split_text(text)``
exactly, with the library defaults: ``keep_separator=True``,
``is_separator_regex=False``, ``strip_whitespace=True``, ``length_function=len``.

Reference: langchain_text_splitters/character.py (RecursiveCharacterTextSplitter)
and langchain_text_splitters/base.py (TextSplitter._merge_splits / _join_docs).
"""

from __future__ import annotations

import re
from typing import List, Optional

DEFAULT_SEPARATORS = ["\n\n", "\n", " ", ""]


def _split_text_with_regex(text: str, separator: str, keep_separator: bool) -> List[str]:
    # Mirrors langchain's _split_text_with_regex with keep_separator truthy
    # (which RecursiveCharacterTextSplitter uses by default == "start").
    if separator:
        if keep_separator:
            # Parentheses keep the delimiters in the result; the "start"
            # behavior attaches each separator to the piece that follows it.
            splits_ = re.split(f"({separator})", text)
            splits = [splits_[i] + splits_[i + 1] for i in range(1, len(splits_), 2)]
            if len(splits_) % 2 == 0:
                splits += splits_[-1:]
            splits = [splits_[0], *splits]
        else:
            splits = re.split(separator, text)
    else:
        splits = list(text)
    return [s for s in splits if s]


def _join_docs(docs: List[str], separator: str, strip_whitespace: bool) -> Optional[str]:
    text = separator.join(docs)
    if strip_whitespace:
        text = text.strip()
    if text == "":
        return None
    return text


def _merge_splits(
    splits,
    separator: str,
    chunk_size: int,
    chunk_overlap: int,
    strip_whitespace: bool = True,
) -> List[str]:
    # Mirrors TextSplitter._merge_splits (length_function == len).
    separator_len = len(separator)

    docs: List[str] = []
    current_doc: List[str] = []
    total = 0
    for d in splits:
        len_ = len(d)
        if total + len_ + (separator_len if len(current_doc) > 0 else 0) > chunk_size:
            # (langchain logs a warning here when total > chunk_size; omitted.)
            if len(current_doc) > 0:
                doc = _join_docs(current_doc, separator, strip_whitespace)
                if doc is not None:
                    docs.append(doc)
                # Pop while we still exceed the overlap, or still don't fit.
                while total > chunk_overlap or (
                    total + len_ + (separator_len if len(current_doc) > 0 else 0)
                    > chunk_size
                    and total > 0
                ):
                    total -= len(current_doc[0]) + (
                        separator_len if len(current_doc) > 1 else 0
                    )
                    current_doc = current_doc[1:]
        current_doc.append(d)
        total += len_ + (separator_len if len(current_doc) > 1 else 0)
    doc = _join_docs(current_doc, separator, strip_whitespace)
    if doc is not None:
        docs.append(doc)
    return docs


def _split_text(
    text: str,
    separators: List[str],
    chunk_size: int,
    chunk_overlap: int,
    keep_separator: bool = True,
    strip_whitespace: bool = True,
) -> List[str]:
    # Mirrors RecursiveCharacterTextSplitter._split_text (is_separator_regex=False).
    final_chunks: List[str] = []
    # Get appropriate separator to use.
    separator = separators[-1]
    new_separators: List[str] = []
    for i, s_ in enumerate(separators):
        separator_ = re.escape(s_)
        if s_ == "":
            separator = s_
            break
        if re.search(separator_, text):
            separator = s_
            new_separators = separators[i + 1:]
            break

    separator_ = re.escape(separator)
    splits = _split_text_with_regex(text, separator_, keep_separator)

    # Now go merging things, recursively splitting longer texts.
    good_splits: List[str] = []
    merge_separator = "" if keep_separator else separator
    for s in splits:
        if len(s) < chunk_size:
            good_splits.append(s)
        else:
            if good_splits:
                merged_text = _merge_splits(
                    good_splits, merge_separator, chunk_size, chunk_overlap,
                    strip_whitespace,
                )
                final_chunks.extend(merged_text)
                good_splits = []
            if not new_separators:
                final_chunks.append(s)
            else:
                other_info = _split_text(
                    s, new_separators, chunk_size, chunk_overlap,
                    keep_separator, strip_whitespace,
                )
                final_chunks.extend(other_info)
    if good_splits:
        merged_text = _merge_splits(
            good_splits, merge_separator, chunk_size, chunk_overlap,
            strip_whitespace,
        )
        final_chunks.extend(merged_text)
    return final_chunks


def recursive_split(text, chunk_size=1000, overlap=150, separators=None):
    """Vendored from langchain RecursiveCharacterTextSplitter (MIT). Returns list[str]."""
    if separators is None:
        separators = list(DEFAULT_SEPARATORS)
    if overlap > chunk_size:
        raise ValueError(
            f"Got a larger chunk overlap ({overlap}) than chunk size "
            f"({chunk_size}), should be smaller."
        )
    return _split_text(text, separators, chunk_size, overlap)
